src/Tools/BuildValidator/CompilationDiff.cs

```diff
@@ -165,9 +176,29 @@ public string MiscErrorMessage
                         assemblyInfo,
                         RebuildResult.BinaryDifference,
                         originalPortableExecutableBytes: originalBytes,
+                        originalPdbReader: compilationFactory.OptionsReader.PdbReader,
                         rebuildPortableExecutableBytes: rebuildBytes,
+                        rebuildPdbReader: getRebuildPdbReader(),
                         rebuildCompilation: rebuildCompilation);
                 }
+
+                unsafe MetadataReader getRebuildPdbReader()
+                {
+                    if (hasEmbeddedPdb)
+                    {
+                        fixed (byte* rebuildPtr = rebuildBytes)
+                        {
+                            var peReader = new PEReader(rebuildPtr, rebuildBytes.Length);
+                            return peReader.GetEmbeddedPdbMetadataReader() ?? throw ExceptionUtilities.Unreachable;
+                        }
+                    }
+                    else
+                    {
+                        // TODO: it appears that readers created from this stream have non-deterministic issues.
+                        // For example, "read out of bounds" exceptions, "method is not supported" exceptions in DiaSymReader, etc.
+                        return MetadataReaderProvider.FromPortablePdbImage(rebuildPdbStream!.ToImmutable()).GetMetadataReader();
```

Comment:
I haven't been able to figure out what's wrong here. I've found that when I force a failure (e.g. by uncommenting the line earlier in this change which forces the optimization level of the rebuild to Release), there are nondeterministic errors when visualizing the PDB.

Sometimes PDB2XML fails, sometimes MDV fails, and the messages vary a bit. So I am wondering if I am messing up creating/using the PDB stream, or... what?

Reply:
Looks like the root cause was actually the following series of events:

2. use TryGetAssocatedPortablePdb to get a MetadataReaderProvider from a FileStream from a portable .pdb
3. Dispose the MetadataReaderProvider
4. use the MetadataReader owned by the provider.

Have changed the impl to prevent this.